staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -254,13 +301,34 @@ func logWatchChannelErr(err error) {
 // - get current objects if initialRev=0; set initialRev to current rev
 // - watch on given key and send events to process.
 func (wc *watchChan) startWatching(watchClosedCh chan struct{}) {
-	if wc.initialRev == 0 {
+	// since we start watching the other goroutine cannot
+	// receive any bookmark events thus it's safe to read the value once
+	initialEventsEndBookmarkSent := wc.isInitialEventsEndBookmarkSent()
+	listObjects := wc.initialRev == 0
+	if wc.initialRev > 0 && !initialEventsEndBookmarkSent {
+		currentStorageRV, err := wc.getCurrentResourceVersionFromStorage()
+		if err != nil {
+			err = fmt.Errorf("failed to get the current resource version from the storage: %w", err)
+			wc.sendError(err)
+			return
+		}
+		if uint64(wc.initialRev) > currentStorageRV {
+			// TODO: add some jitter to resourceVersionTooHighRetrySeconds
+			wc.sendError(storage.NewTooLargeResourceVersionError(uint64(wc.initialRev), currentStorageRV, resourceVersionTooHighRetrySeconds))
+			return
+		}
+		listObjects = true
+	}
+	if listObjects {
 		if err := wc.sync(); err != nil {
 			klog.Errorf("failed to sync with latest state: %v", err)
 			wc.sendError(err)
 			return
 		}
 	}
+	if !initialEventsEndBookmarkSent {
+		wc.sendEvent(progressNotifyEvent(wc.initialRev))
```

Comment:
we could extent the `event` struct and pass information that this bookmark needs to be annotated.
that would allows us to get rid of the lock.

Reply:
Looking at those changes here, I think it would be much simpler [a lot of this code wouldn't even be needed]

So yeah - let's go that way.

Reply:
ok, thanks